src/ipa_pc/mod.rs

```diff
@@ -560,24 +767,102 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
                 "labeled_comm degree bound mismatch for {}",
                 label
             );
-            if let Some(degree_bound) = degree_bound {
-                let shifted_polynomial = Self::shift_polynomial(ck, polynomial, degree_bound);
-                combined_polynomial += (cur_challenge, &shifted_polynomial);
-                combined_commitment_proj += &commitment.shifted_comm.unwrap().mul(cur_challenge);
 
-                if hiding_bound.is_some() {
-                    let shifted_rand = randomness.shifted_rand;
-                    assert!(
-                        shifted_rand.is_some(),
-                        "shifted_rand.is_none() for {}",
-                        label
+            if hiding_bound.is_some() {
+                has_hiding = true;
+            }
+
+            if segments_count > 1 {
+
+                let mut polynomial_lc = Polynomial::zero();
+                let mut comm_lc = <G::Projective as ProjectiveCurve>::zero();
+                let mut rand_lc = G::ScalarField::zero();
+
+                for i in 0..segments_count {
+                    let is = i * key_len;
+                    let poly_single = Polynomial::from_coefficients_slice(
+                        &polynomial.coeffs[i * key_len..core::cmp::min((i + 1) * key_len, p_len)]
                     );
-                    combined_rand += &(cur_challenge * &shifted_rand.unwrap());
+                    let comm_single = commitment.comm[i];
+                    polynomial_lc += (point.pow(&[is as u64]), &poly_single);
+                    comm_lc += &comm_single.mul(point.pow(&[is as u64]));
+                    if has_hiding {
+                        let rand_single = randomness.rand[i];
+                        rand_lc += &(point.pow(&[is as u64]) * rand_single);
+                    }
+                }
+
+                combined_polynomial += (cur_challenge, &polynomial_lc);
+                combined_commitment_proj += &comm_lc.mul(&cur_challenge);
+                if has_hiding {
+                    combined_rand += &(cur_challenge * &rand_lc);
+                }
+
+            } else {
+
+                combined_polynomial += (cur_challenge, polynomial);
+                combined_commitment_proj += &commitment.comm[0].mul(cur_challenge);
+
+                if has_hiding {
+                    combined_rand += &(cur_challenge * &randomness.rand[0]);
                 }
             }
 
             cur_challenge = opening_challenges(opening_challenge_counter);
             opening_challenge_counter += 1;
+
+            if let Some(degree_bound_len) = degree_bound_len {
+
+                if segments_count > 1 {
+
+                    let shifted_degree_bound = degree_bound_len % key_len - 1;
```

Comment:
This formula does not capture the corner case `degree_bound_len % key_len == 0``.
I think we need to set `shifted_degree_bound = key_len - 1 ` in case that `degree_bound_len % key_len == 0`.

Reply:
This corner key is always checked some lines of code before with this:

let degree_bound_len = degree_bound.and_then(|degree_bound| {
    if (degree_bound + 1) % key_len != 0 { Some(degree_bound + 1) } else { None }
});

So in case degree_bound_len % key_len == 0 we never reached this part of code